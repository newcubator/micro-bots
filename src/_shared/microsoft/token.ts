import axios from "axios";
import { stringify } from "node:querystring";

// Cache
let cachedAccessToken: string = undefined;
let cachedAccessTokenExpiration: number = undefined;

export function getAzureAccessToken(): Promise<string> {
  const MICROSOFT_TOKEN = process.env.MICROSOFT_TOKEN;
  if (!MICROSOFT_TOKEN) {
    return Promise.reject(new Error("Microsoft graph api token configuration missing"));
  }

  const MICROSOFT_CLIENT_ID = process.env.MICROSOFT_CLIENT_ID;
  if (!MICROSOFT_CLIENT_ID) {
    return Promise.reject(new Error("Microsoft graph api client id configuration missing"));
  }

  const MICROSOFT_CLIENT_SECRET = process.env.MICROSOFT_CLIENT_SECRET;
  if (!MICROSOFT_CLIENT_SECRET) {
    return Promise.reject(new Error("Microsoft graph api client secret configuration missing"));
  }

  if (cachedAccessToken && Date.now() < cachedAccessTokenExpiration) {
    return Promise.resolve(cachedAccessToken);
  }

  console.info("Requesting microsoft graph api access token.");
  return axios
    .post(
      `https://login.microsoftonline.com/${MICROSOFT_TOKEN}/oauth2/v2.0/token`,
      stringify({
        client_id: MICROSOFT_CLIENT_ID,
        client_secret: MICROSOFT_CLIENT_SECRET,
        scope: "https://graph.microsoft.com/.default",
        grant_type: "client_credentials",
      }),
      {
        headers: { "Content-Type": `application/x-www-form-urlencoded` },
      },
    )
    .then((response) => {
      // update cache
      cachedAccessToken = response.data.access_token;
      cachedAccessTokenExpiration = Date.now() + toMilliSeconds(response.data.expires_in - 20);

      return cachedAccessToken;
    })
    .catch((error) => {
      console.error("Error while calling microsoft graph api to get access token!");
      throw error;
    });
}

function toMilliSeconds(timeInSeconds: number): number {
  return timeInSeconds * 1000;
}
