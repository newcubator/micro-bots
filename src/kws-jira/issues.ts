import axios from "axios";
import { Issue, SprintIssues } from "./types/jira-types";
import { JIRA_USER_KWS, JIRA_PASSWORD_KWS } from "./credentials";

export async function getIssue(key: string): Promise<Issue | null> {
  const url = new URL("https://kwssaat.atlassian.net/rest/api/2/search");
  url.searchParams.append("jql", `issuekey= "${key}"`);

  const response = await axios.get<SprintIssues>(url.toString(), {
    auth: {
      username: JIRA_USER_KWS,
      password: JIRA_PASSWORD_KWS,
    },
  });
  const sprint = response.data;

  return sprint.issues && sprint.issues.length > 0 ? sprint.issues[0] : null;
}

export async function getIssues(keys: string[], maxResults = 100): Promise<Issue[]> {
  const jql = `key in (${keys.join(",")})`;
  let startAt = 0;
  let allIssues: Issue[] = [];
  let issues: Issue[] = [];

  do {
    const url = new URL("https://kwssaat.atlassian.net/rest/api/2/search");
    url.searchParams.append("jql", jql);
    url.searchParams.append("startAt", String(startAt));
    url.searchParams.append("maxResults", String(maxResults));
    url.searchParams.append(
      "fields",
      "issuetype,created,description,summary,customfield_10027,customfield_10089,customfield_10010,timeestimate,timetracking",
    );

    const response = await axios.get<SprintIssues>(url.toString(), {
      auth: {
        username: JIRA_USER_KWS,
        password: JIRA_PASSWORD_KWS,
      },
    });

    const sprint = response.data;
    issues = sprint.issues || [];

    allIssues = [...allIssues, ...issues];

    startAt += maxResults;
  } while (issues.length === maxResults);

  return allIssues;
}
