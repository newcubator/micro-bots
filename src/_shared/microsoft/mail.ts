import axios from "axios";
import dayjs from "dayjs";
import { getAzureAccessToken } from "./token";

export async function getMessage(userPrincipalName: string, messageId: string): Promise<MicrosoftMessage> {
  const accessToken = await getAzureAccessToken();

  return axios
    .get(`https://graph.microsoft.com/v1.0/users/${userPrincipalName}/messages/${messageId}`, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
        Prefer: 'IdType="ImmutableId",outlook.body-content-type="text"',
      },
    })
    .then((response) => {
      console.debug(response.data);
      return response.data;
    })
    .catch((error) => {
      console.error("Error while calling microsoft graph api to get message!");
      throw error;
    });
}

export async function getMessageAttachments(
  userPrincipalName: string,
  messageId: string
): Promise<MicrosoftAttachment[]> {
  const accessToken = await getAzureAccessToken();

  return axios
    .get(`https://graph.microsoft.com/v1.0/users/${userPrincipalName}/messages/${messageId}/attachments`, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
    })
    .then((response) => {
      console.debug(response.data.value);
      return response.data.value;
    })
    .catch((error) => {
      console.error("Error while calling microsoft graph api to get message attachments!");
      throw error;
    });
}

export async function markMessageAsRead(userPrincipalName: string, messageId: string) {
  const accessToken = await getAzureAccessToken();

  return axios
    .patch(
      `https://graph.microsoft.com/v1.0/users/${userPrincipalName}/messages/${messageId}`,
      {
        isRead: true,
      },
      {
        headers: {
          Authorization: `Bearer ${accessToken}`,
        },
      }
    )
    .then((response) => {
      console.debug(response.data);
      return;
    })
    .catch((error) => {
      console.error("Error while calling microsoft graph api to mark message as read!");
      throw error;
    });
}

export async function moveMessage(userPrincipalName: string, messageId: string, destinationId: string) {
  const accessToken = await getAzureAccessToken();

  return axios
    .post(
      `https://graph.microsoft.com/v1.0/users/${userPrincipalName}/messages/${messageId}/move`,
      {
        destinationId,
      },
      {
        headers: {
          Authorization: `Bearer ${accessToken}`,
        },
      }
    )
    .then((response) => {
      console.debug(response.data);
      return;
    })
    .catch((error) => {
      console.error("Error while calling microsoft graph api to move message!");
      throw error;
    });
}

export async function changeMailRespondForUser(microsoftUserId, message, startDate, endDate) {
  const accessToken = await getAzureAccessToken();

  const mailData = {
    "@odata.context": `https://graph.microsoft.com/v1.0/$metadata#users('${microsoftUserId}')/mailboxSettings`,
    automaticRepliesSetting: {
      status: "scheduled",
      externalAudience: "all",
      internalReplyMessage: message,
      scheduledStartDateTime: {
        dateTime: dayjs(startDate).utc().format(),
        timeZone: "UTC",
      },
      scheduledEndDateTime: {
        dateTime: dayjs(endDate).add(1, "day").utc().format(),
        timeZone: "UTC",
      },
    },
  };

  return axios
    .patch(`https://graph.microsoft.com/v1.0/users/${microsoftUserId}/mailboxSettings`, mailData, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
    })
    .then((response) => {
      return response.data;
    })
    .catch((error) => {
      console.log(error);
    });
}

export async function getMailSettingsForUser(microsoftUserId) {
  const accessToken = await getAzureAccessToken();

  return axios
    .get(`https://graph.microsoft.com/v1.0/users/${microsoftUserId}/mailboxSettings`, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
    })
    .then((response) => {
      return response.data;
    })
    .catch((error) => {
      console.log(error);
    });
}

/**
 * Message Resource
 *
 * Only used properties are mapped!
 * @See https://learn.microsoft.com/en-us/graph/api/resources/message?view=graph-rest-1.0#properties
 */
export interface MicrosoftMessage {
  id: string;
  from: {
    emailAddress: {
      address: string;
      name: string;
    };
  };
  subject: string;
  body: {
    content: string;
  };
  hasAttachments: boolean;
  webLink: string;
}

/**
 * Mail Attachment Resource
 *
 * Only used properties are mapped!
 * @See https://learn.microsoft.com/en-us/graph/api/resources/fileattachment?view=graph-rest-1.0#properties
 */
export interface MicrosoftAttachment {
  contentId: string;
  contentBytes: string;
  contentType: string;
  name: string;
  isInline: boolean;
}
