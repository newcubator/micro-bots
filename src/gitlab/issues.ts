import axios from "axios";
import { GitlabIssue } from "./gitlab";
import { GITLAB_TOKEN } from "./token";

export const getIssues = (projectId: string, search?: string, searchIn?: "title" | "description") => {
  return axios.get<GitlabIssue[]>(`https://gitlab.com/api/v4/projects/${projectId}/issues`, {
    headers: {
      Authorization: `Bearer ${GITLAB_TOKEN}`,
    },
    params: {
      search,
      in: searchIn,
    },
  });
};

export const postIssue = (
  projectId: string,
  title: string,
  description?: string,
  labels?: string[],
  due_date?: string
) => {
  return axios.post<GitlabIssue>(
    `https://gitlab.com/api/v4/projects/${projectId}/issues`,
    {
      title,
      description,
      labels,
      due_date,
    } as GitlabIssue,
    {
      headers: {
        Authorization: `Bearer ${GITLAB_TOKEN}`,
      },
    }
  );
};
