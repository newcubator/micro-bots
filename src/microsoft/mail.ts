import axios from "axios";
import dayjs from "dayjs";
import { getAzureAccessToken } from "./token";
import utc from "dayjs/plugin/utc";

dayjs.extend(utc);

export async function changeMailRespondForUser(microsoftUserId, startDate, endDate) {
  const azureAccessToken = await getAzureAccessToken();

  startDate = dayjs(startDate).utc().format();
  endDate = dayjs(endDate).utc().format();

  const mailData = {
    "@odata.context": `https://graph.microsoft.com/v1.0/$metadata#users('${microsoftUserId}')/mailboxSettings`,
    automaticRepliesSetting: {
      status: "scheduled",
      externalAudience: "all",
      internalReplyMessage:
        '<html><body><div style="font-family:Calibri,Arial,Helvetica,sans-serif; font-size:12pt; color:rgb(0,0,0)">Hier steht auch eine Testnachricht f√ºr die Kollegen drin!!</div></body></html>',
      externalReplyMessage:
        '<html><body><div style="font-family:Calibri,Arial,Helvetica,sans-serif; font-size:12pt; color:rgb(0,0,0)">Hier Steht eine Testnachricht an Fremde Menschen drin!!</div></body></html>',
      scheduledStartDateTime: {
        dateTime: startDate,
        timeZone: "UTC",
      },
      scheduledEndDateTime: {
        dateTime: endDate,
        timeZone: "UTC",
      },
    },
  };

  return axios
    .patch(`https://graph.microsoft.com/v1.0/users/${microsoftUserId}/mailboxSettings`, mailData, {
      headers: { Authorization: `Bearer ${azureAccessToken.access_token}` },
    })
    .then((response) => {
      console.log("ResponseData", response.data);
      return response.data;
    })
    .catch((error) => {
      console.log(error);
    });
}
