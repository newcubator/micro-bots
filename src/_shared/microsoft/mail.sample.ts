import { MicrosoftAttachment, MicrosoftMessage } from "./mail";

export const SAMPLE_MESSAGE: MicrosoftMessage = {
  id: "AQMkADQ3YTAyOWE2LTdjNjAtNDhjMy1iNWY4LTY0Y2RjODI0MmE0MABGAAAVc",
  from: {
    emailAddress: {
      address: "applicant@jestjs.io",
      name: "Job Applicant",
    },
  },
  subject: "Job Application",
  body: {
    content: "I would like to work for you",
  },
  hasAttachments: true,
  webLink: "https://outlook.office.com/mail/abc123",
};

export const SAMPLE_ATTACHMENT_1: MicrosoftAttachment = {
  contentId: "BD9DDB4C-2B48-4A5B-8F4D-D8327DCCC23D",
  name: "ori3.jpg",
  contentBytes: "some jpeg bytes ...",
  contentType: "image/jpeg",
  isInline: true,
};

export const SAMPLE_ATTACHMENT_2: MicrosoftAttachment = {
  contentId: "F2183E82104CCA4FB7A3E3B4C54F1BD1@EURP194.PROD.OUTLOOK.COM",
  name: "Lebenslauf.pdf",
  contentBytes: "some pdf bytes ...",
  contentType: "application/octet-stream",
  isInline: false,
};

export const SAMPLE_ATTACHMENTS: MicrosoftAttachment[] = [SAMPLE_ATTACHMENT_1, SAMPLE_ATTACHMENT_2];
