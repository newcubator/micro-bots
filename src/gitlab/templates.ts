import axios from 'axios';
import { GitlabIssueTemplate } from './gitlab';
import { GITLAB_TOKEN } from './token';

export const getIssueTemplateByName = async (projectId: string, name: string) => {
  return await axios.get<GitlabIssueTemplate>(encodeURI(`https://gitlab.com/api/v4/projects/${projectId}/templates/issues/${name}`), {
    headers: {
      Authorization: `Bearer ${GITLAB_TOKEN}`,
    },
  });
};
