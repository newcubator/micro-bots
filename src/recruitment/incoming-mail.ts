import { APIGatewayRequestAuthorizerEventV2 } from "aws-lambda";
import { getMessage, getMessageAttachments, markMessageAsRead, moveMessage } from "../_shared/microsoft/mail";
import { GitlabIssue } from "../gitlab/gitlab";
import { getIssueTemplateByName } from "../gitlab/templates";
import { createIssue, updateIssueDescription, uploadFile } from "../gitlab/issues";

const GITLAB_PROJECT_ID = "35095396";
const TEMPLATE_NAME = "Bewerbung";
const MAIL_USER_PRINCIPAL_NAME = "info@newcubator.com";
const MAIL_ARCHIVE_FOLDER_ID =
  "AAMkAGJmODcwNWExLWM3ZmItNGNlZS04N2Q2LTc4YTJkMTZkOGJmOQAuAAAAAACsW_joJTKFSIXU5SXQHv11AQA4LQp9ObtJT7LiE4tqoOmkAAOkGk5ZAAA=";
const MAIL_LINK_TEMPLATE = `
<b><a href="mailto:[RECIPIENT]?bcc=incoming+newcubator-people-employees-35095396-dxmv8fhkpi28plc4lptak2z5l-issue-[IID]@incoming.gitlab.com"> >> Mail senden und als Kommentar anh√§ngen</a></b>
`;

export const handler = async (event: APIGatewayRequestAuthorizerEventV2) => {
  let messageId = event.queryStringParameters.id;
  console.log("messageId", messageId);

  if (!messageId) {
    console.error("No messageId found");
    return {
      statusCode: 400,
      body: "Missing 'messageId'",
    };
  }

  const [message, attachments, template] = await Promise.all([
    getMessage(MAIL_USER_PRINCIPAL_NAME, messageId),
    getMessageAttachments(MAIL_USER_PRINCIPAL_NAME, messageId),
    getIssueTemplateByName(GITLAB_PROJECT_ID, TEMPLATE_NAME),
  ]);

  // update id with immutable id
  messageId = message.id;

  let description = template.content;
  description = description.replace("[NAME]", message.from.emailAddress.name);
  description = description.replace("[EMAIL]", message.from.emailAddress.address);
  description = description.replace("[BODY]", message.body.content);
  description = description.replace("[WEBLINK]", message.webLink);

  await Promise.all(
    attachments.map(async (attachment) => {
      const upload = await uploadFile(
        GITLAB_PROJECT_ID,
        attachment.contentBytes,
        attachment.contentType,
        attachment.name
      );
      if (attachment.isInline) {
        description = description.replace(`[cid:${attachment.contentId}]`, upload.markdown);
      } else {
        description += `\r\n${upload.markdown}`;
      }
    })
  );

  const issue = await createIssue(GITLAB_PROJECT_ID, message.subject, description);

  await Promise.all([
    updateIssueWithMailLink(issue, message.from.emailAddress.address),
    markMessageAsRead(MAIL_USER_PRINCIPAL_NAME, messageId),
    moveMessage(MAIL_USER_PRINCIPAL_NAME, messageId, MAIL_ARCHIVE_FOLDER_ID),
  ]);

  return {
    statusCode: 200,
    body: JSON.stringify({
      result: "Done",
    }),
  };
};

async function updateIssueWithMailLink(issue: GitlabIssue, recipient: string) {
  let link = MAIL_LINK_TEMPLATE;
  link = link.replace("[IID]", issue.iid.toString());
  link = link.replace("[RECIPIENT]", recipient);
  return updateIssueDescription(GITLAB_PROJECT_ID, issue.iid, link + issue.description);
}
