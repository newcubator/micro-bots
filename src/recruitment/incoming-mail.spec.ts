import { APIGatewayRequestAuthorizerEventV2 } from "aws-lambda";
import { SAMPLE_ATTACHMENTS, SAMPLE_MESSAGE } from "../_shared/microsoft/mail.sample";
import { handler } from "./incoming-mail";
import { getMessageAttachments, getMessage, markMessageAsRead, moveMessage } from "../_shared/microsoft/mail";
import { getIssueTemplateByName } from "../gitlab/templates";
import { createIssue, updateIssueDescription, uploadFile } from "../gitlab/issues";

jest.mock("../_shared/microsoft/mail");
jest.mock("../gitlab/templates");
jest.mock("../gitlab/issues");

const getMessageMock = getMessage as unknown as jest.Mock;
const getIssueTemplateMock = getIssueTemplateByName as jest.Mock;
const getAttachmentsMock = getMessageAttachments as jest.Mock;
const uploadFileMock = uploadFile as jest.Mock;
const createIssueMock = createIssue as jest.Mock;
const updateIssueDescriptionMock = updateIssueDescription as jest.Mock;
const markMessageAsReadMock = markMessageAsRead as jest.Mock;
const moveMessageMock = moveMessage as jest.Mock;

describe("Recruitment Incoming Job Application Mail", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should create gitlab issue from mail job application", async () => {
    getMessageMock.mockResolvedValueOnce(SAMPLE_MESSAGE);
    getIssueTemplateMock.mockResolvedValueOnce(SAMPLE_TEMPLATE);
    getAttachmentsMock.mockResolvedValueOnce(SAMPLE_ATTACHMENTS);
    uploadFileMock.mockResolvedValueOnce({
      markdown: "![ori3.jpg](https://gitlab.com/ori3.jpg)",
    });
    uploadFileMock.mockResolvedValueOnce({
      markdown: "![Lebenslauf.pdf](https://gitlab.com/Lebenslauf.pdf)",
    });
    createIssueMock.mockResolvedValue({
      id: 13,
      iid: 6,
    });
    markMessageAsReadMock.mockResolvedValueOnce({});

    await expect(handler(SAMPLE_EVENT)).resolves.toMatchObject({
      statusCode: 200,
      body: '{"result":"Done"}',
    });

    expect(createIssueMock.mock.calls[0]).toMatchSnapshot();
    expect(updateIssueDescriptionMock.mock.calls[0]).toMatchSnapshot();
    expect(markMessageAsReadMock).toHaveBeenCalledWith(
      "info@newcubator.com",
      "AQMkADQ3YTAyOWE2LTdjNjAtNDhjMy1iNWY4LTY0Y2RjODI0MmE0MABGAAAVc"
    );
    expect(moveMessageMock).toHaveBeenCalledWith(
      "info@newcubator.com",
      "AQMkADQ3YTAyOWE2LTdjNjAtNDhjMy1iNWY4LTY0Y2RjODI0MmE0MABGAAAVc",
      "AAMkAGJmODcwNWExLWM3ZmItNGNlZS04N2Q2LTc4YTJkMTZkOGJmOQAuAAAAAACsW_joJTKFSIXU5SXQHv11AQA4LQp9ObtJT7LiE4tqoOmkAAOkGk5ZAAA="
    );
  });

  it("should return error when mail messageId is missing", async () => {
    const event = {
      queryStringParameters: {
        id: undefined,
      },
    } as any as APIGatewayRequestAuthorizerEventV2;

    await expect(handler(event)).resolves.toMatchObject({
      statusCode: 400,
      body: "Missing 'messageId'",
    });

    expect(getMessageMock).toBeCalledTimes(0);
  });
});

const SAMPLE_EVENT = {
  queryStringParameters: {
    id: "AQMkADQ3YTAyOWE2LTdjNjAtNDhjMy1iNWY4LTY0Y2RjODI0MmE0MABGAAAVc",
  },
} as any as APIGatewayRequestAuthorizerEventV2;

const SAMPLE_TEMPLATE = {
  content: `Template Content
[NAME]
[EMAIL]
[BODY]
[WEBLINK]`,
};
