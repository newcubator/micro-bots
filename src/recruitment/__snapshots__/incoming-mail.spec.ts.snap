// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recruitment Incoming Job Application Mail should create gitlab issue from mail job application 1`] = `
[
  "35095396",
  "Job Application",
  "Template Content
Job Applicant
applicant@jestjs.io
I would like to work for you
https://outlook.office.com/mail/abc123
![Lebenslauf.pdf](https://gitlab.com/Lebenslauf.pdf)",
]
`;

exports[`Recruitment Incoming Job Application Mail should create gitlab issue from mail job application 2`] = `
[
  "35095396",
  6,
  "
<b><a href="mailto:applicant@jestjs.io?bcc=incoming+newcubator-people-employees-35095396-dxmv8fhkpi28plc4lptak2z5l-issue-6@incoming.gitlab.com"> >> Mail senden und als Kommentar anh√§ngen</a></b>
undefined",
]
`;
